### PR TITLE
SDS-430: add validateBeforeTouched option to ReduxFormTextInput

### PR DESCRIPTION
**Pull request by burrata**
#### Related issues
Fixes https://meetup.atlassian.net/browse/SDS-430

#### Description
For field-level validations, redux-form automatically validates on first render unless you check for validation on whether the field has been touched.

It seems as though this is intended behavior.
https://github.com/erikras/redux-form/issues/120
>This is the purpose of the touched flags. In most UI cases, you should only show an error if a field has been touched (gone to and then blurred). There is also a visited flag, which is set to true onFocus. There's a working example of how the flags work here.
>Does that help?

Docs for `meta.touched`: https://redux-form.com/7.0.4/docs/api/field.md/#-code-meta-touched-boolean-code-

This ticket is to adjust the current `ReduxFormTextInput` implementation to take a flag that, when passed in, does not show the initial validation error until after the field has been touched. This is to prevent having to check for this in all form implementations we have.

#### Screenshots (if applicable)
Add venue form on first load, with these fields being required:
screen shot 2017-10-04 at 5 04 30 pm

### CODE REVIEW

File: src/forms/redux-form/textInput.test.jsx [hunk 1]
```diff
@@ -13,15 +14,47 @@ describe('redux-form TextInput', function() {
...
+	});
 
+	describe('validateAfterTouched', () => {
+		it('always renders the error when `validateAfterTouched` is not true', () => {
+			const component = shallow(<ReduxFormTextInput {...formAttrs} />);
+			expect(component.prop('error')).toBe(MOCK_ERROR);
+		});
+		describe('when true', () => {
+			it('does not render the error when touched is false', () => {
+				const props = {
```
Comment (unjust) [inline src/forms/redux-form/textInput.test.jsx]: just pull these props out and set `props.validateAfterTouched = true` or `false` in each test

**Comment by unjust** [comment]
I wonder if there are other ways to prevent this behavior rather than adding into `TextInput`. (I'm hesitant to embed an extra property here.)

Right now, in the event schedule form I think we don't see this issue because we avoid it - by not checking required fields in the [`validate`](https://redux-form.com/7.0.4/docs/api/reduxform.md/#-code-validate-values-object-props-object-gt-errors-object-code-optional-) function of the form.

Maybe we can look into leveraging the `initialRender` property of 
`shouldValidate` instead so when `validate` gets called we skip certain fields. https://redux-form.com/7.0.4/docs/api/reduxform.md/#-code-shouldvalidate-params-boolean-code-optional-

@mikespencer have you seen this issue with your forms on pro? how have you avoided it

**Comment by mperrotti** [comment]
@mikespencer - We follow semver

**Comment by mikespencer** [comment]
> @mikespencer have you seen this issue with your forms on pro? how have you avoided it

@unjust We've been using form components (eg [Input](https://github.com/meetup/pro-web/blob/master/src/app/shared/components/fields/Input.jsx)) that we created a while ago in pro-web before redux-form compatible form components were available here. We will need to update this on our end eventually, just haven't got around to it yet. These components use the `touched && error` check, though.

Seems a common pattern elsewhere too: [redux-form-material-ui abstract this logic into a reusable function](https://github.com/erikras/redux-form-material-ui/blob/master/src/mapError.js#L9)

**Comment by burrata** [comment]
@unjust are you good with this change, per @mikespencer's thoughts above?

@mperrotti should I update the semver if I merge?

File: src/forms/redux-form/TextInput.jsx [hunk 2]
```diff
@@ -8,9 +8,10 @@ import TextInput from '../TextInput';
  * @return {Component} TextInput
  */
 const ReduxFormTextInput = props => {
-	const { meta, input, ...other } = props;
+	const { meta, input, validateAfterTouched, ...other } = props;
+	const error = (!validateAfterTouched || meta.touched) ? meta.error : null;
```
Comment (unjust) [inline src/forms/redux-form/TextInput.jsx]: Should this be `validateBeforeTouched` instead? Just thinking what the most common use case might be so we don't have to add an extra prop every time. I would think we would in most cases, want always to check `touched` before validating.

File: src/forms/redux-form/textInput.test.jsx (hunk 1 above)
Comment (unjust) [inline src/forms/redux-form/textInput.test.jsx]: just pull these props out and set `props.meta.touched = true` or `false` in each test. I htink its more legible whats going on but you can disagree.

File: src/forms/redux-form/TextInput.jsx (hunk 2 above)
Comment (unjust) [reply src/forms/redux-form/TextInput.jsx]: also if we can add some documentation somewhere as to what this property is there for, that would be 💯

Comment (mikespencer) [reply src/forms/redux-form/TextInput.jsx]: > Should this be validateBeforeTouched instead?

Actually I think that makes more sense too 👍

Comment (burrata) [reply src/forms/redux-form/TextInput.jsx]: I'll make this update to both my prs!

**Comment by unjust** [comment]
I approve. Ready to moige.

**Pull request merged**